### PR TITLE
CORDA-3048: Add java.util.Currency support.

### PR DESCRIPTION
**Pull request by chrisr3**
The Tokens SDK needs to be able to load `java.util.Currency` objects along with their basic properties. Some of this information is loaded from a `currency.data` file while other information is loaded from the `CurrencyNames` resource bundle.

For now, just support the built-in fragment of the `CurrencyNames` resource bundle, i.e. English language only.

This also allows sandboxed code to execute `Class.netInstance()` and `Constructor.newInstance()` now for the sake of the Kotlin standard libraries.

### CODE REVIEW

File: djvm/src/main/java/sandbox/java/util/Currency.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    static int[] scOldCurrenciesNumericCode;
+    static int[] scNewCurrenciesNumericCode;
+    static String otherCurrencies;
+    static int[] otherCurrenciesDFD;
+    static int[] otherCurrenciesNumericCode;
+
+    static {
+        AccessController.doPrivileged(new PrivilegedAction<Void>() {
+            @Override
+            public Void run() {
```
Comment (IgorNitto) [inline djvm/src/main/java/sandbox/java/util/Currency.java]: Same point as in previous PR, a _dummy_ class is equivalent to a compiler stub by definition. If it has any logic into it then it is *not* a dummy class. It is not pedantry: how can i guess what is the logic in this constructor relevant for?

File: djvm/src/main/kotlin/sandbox/java/lang/DJVM.kt [hunk 2]
```diff
@@ -241,6 +242,13 @@ fun getClassLoader(type: Class<*>): ClassLoader {
     return systemClassLoader
 }
 
+/**
+ * Replacement function for [ClassLoader.getSystemResourceAsStream].
+ */
+fun getSystemResourceAsStream(name: kotlin.String): InputStream? {
```
Comment (IgorNitto) [inline djvm/src/main/kotlin/sandbox/java/lang/DJVM.kt]: Where is this function used other than in the "dummy" Currency.java?

File: djvm/src/main/kotlin/net/corda/djvm/rules/implementation/ReturnTypeWrapper.kt [hunk 3]
```diff
@@ -1,27 +1,73 @@
...
+import org.objectweb.asm.Opcodes.INVOKESTATIC
+import org.objectweb.asm.Opcodes.INVOKEVIRTUAL
 
 /**
  * Whitelisted classes may still return [java.lang.String] from some
  * functions, e.g. [java.lang.Object.toString]. So always explicitly
  * invoke [sandbox.java.lang.String.toDJVM] after these.
+ *
+ * These factory functions also need special handling:
+ *   [java.util.concurrent.atomic.AtomicIntegerFieldUpdater.newUpdater]
```
Comment (IgorNitto) [inline djvm/src/main/kotlin/net/corda/djvm/rules/implementation/ReturnTypeWrapper.kt]: What is special about `AtomicIntegerFieldUpdater`? I am missing the connection between this class and `Currency`

File: djvm/src/main/kotlin/net/corda/djvm/analysis/AnalysisConfiguration.kt [hunk 4]
```diff
@@ -384,19 +398,177 @@ class AnalysisConfiguration private constructor(
...
+        ).mapByClassName() + listOf(
+            // Create factory function to wrap java.io.InputStream.
+            object : MethodBuilder(
+                access = ACC_PUBLIC or ACC_STATIC,
+                className = sandboxed(InputStream::class.java),
+                memberName = "toDJVM",
+                descriptor = "(Ljava/io/InputStream;)Lsandbox/java/io/InputStream;"
+            ) {
+                override fun writeBody(emitter: EmitterModule) = with(emitter) {
+                    val doWrap = Label()
```
Comment (IgorNitto) [inline djvm/src/main/kotlin/net/corda/djvm/analysis/AnalysisConfiguration.kt]: I am missing a link here. `Currency` will probably need to load a data file using `FileInputStream`, which will not be possible inside the sandbox. So here we provide `DJVMInputStream` wrapping a generic input stream and a `toDJVM` function, but how is the `Currency` class able to invoke it?

File: djvm/src/main/java/sandbox/java/util/Currency.java (hunk 1 above)
Comment (chrisr3) [reply djvm/src/main/java/sandbox/java/util/Currency.java]: `java.util.Currency` is a dummy class, but `java.util.Currency$1` is _not_:
```java
// Compiles to java.util.Currency$1
new PrivilegedAction<Void>() {
    @Override
    public void run() {
        // etc etc
    }
}
```

File: djvm/src/main/kotlin/sandbox/java/lang/DJVM.kt (hunk 2 above)
Comment (chrisr3) [reply djvm/src/main/kotlin/sandbox/java/lang/DJVM.kt]: It's used in the _non-dummy_ `java.util.Currency$1`.

File: djvm/src/main/kotlin/net/corda/djvm/rules/implementation/ReturnTypeWrapper.kt (hunk 3 above)
Comment (chrisr3) [reply djvm/src/main/kotlin/net/corda/djvm/rules/implementation/ReturnTypeWrapper.kt]: The `AtomicLongFieldUpdater` is used by `Currency`. All _3_ `Atomic*FieldUpdater` classes are used by the Tokens SDK / Kotlin (which was my use-case for testing `Currency`).

File: djvm/src/main/kotlin/net/corda/djvm/analysis/AnalysisConfiguration.kt (hunk 4 above)
Comment (chrisr3) [reply djvm/src/main/kotlin/net/corda/djvm/analysis/AnalysisConfiguration.kt]: The `currency.data` file lives inside `deterministic-rt.jar` and is loaded using `DJVM.getSystemResourceAsStream`.

**Comment by chrisr3** [comment]
> I suspect there are similar patched required to run Java dates and calendar API?
There will need to be _something_, yes. Although [CORDA-1900](https://r3-cev.atlassian.net/browse/CORDA-1900) looks relevant.

**Pull request merged**